tokio/src/sync/task/atomic_waker.rs

```diff
@@ -238,13 +255,26 @@ impl AtomicWaker {
                 debug_assert!(state == REGISTERING || state == REGISTERING | WAKING);
             }
         }
+
+        struct PanicGuard<'a>(&'a AtomicUsize);
+
+        impl Drop for PanicGuard<'_> {
+            fn drop(&mut self) {
+                // On panics, we want to unset the REGISTERING state which will
+                // preserve any potential WAKING state for future calls to
+                // register.
+                let _ = self.0.fetch_and(!REGISTERING, AcqRel);
+            }
+        }
```

Comment:
I don't think this is correct. If the state becomes `WAKING` due to this, then who sets it back to `WAITING`? It seems like the wake lock would just be held indefinitely.

Reply:
Correct, so there's a chance it would get stuck in a `WAKING` state instead due to this.

This is being discussed on Discord, but either revert to old behavior or cope with it somehow. But in the current state at least, no merge.

Reply:
I think that correct behavior would be to:

 1. Replace the current waker with `None`. Store the old waker for now.
 2. Call `waker.into_waker()`.
 3. Set the current waker to `Some(new_waker)`.
 4. Do all of the other stuff in the function.
 5. Just before returning from `do_register` (even if by panic), drop the old waker, if any. Do nothing special if it panics.

In principle this can double-panic in step five, but I do not think that this is a problem since it is normal that things get their destructor called during panics.